release-23.1: upgrade: remove buggy TTL repair

ecwall opened the pull request:
Backport 2/2 commits from #110364 and https://github.com/cockroachdb/cockroach/pull/110816.

/cc @cockroachdb/release

---

Fixes #110363

The TTL descriptor repair in FirstUpgradeFromReleasePrecondition incorrectly
removes TTL fields from table descriptors after incorrectly comparing the
table descriptor's TTL job schedule ID to a set of job IDs.

This change removes the repair until tests are properly added.

Release note (bug fix): Remove buggy TTL descriptor repair. Previously,
upgrading from 22.2.X to 23.1.9 incorrectly removed TTL storage params from
tables (visible via `SHOW CREATE TABLE <ttl-table>;`) while attempting to
repair table descriptors. This resulted in the node that attempts to run the
TTL job crashing due to a panic caused by the missing TTL storage params.
Clusters currently on 22.2.X should NOT be upgraded to 23.1.9 and should
be upgraded to 23.1.10 or later directly.

Release justification: Backport to base extraordinary release 23.1.10 on.

ecwall committed:
upgrade: remove buggy TTL repair

Fixes #110363

The TTL descriptor repair in FirstUpgradeFromReleasePrecondition incorrectly
removes TTL fields from table descriptors after incorrectly comparing the
table descriptor's TTL job schedule ID to a set of job IDs.

This change removes the repair until tests are properly added.

Release note (bug fix): Remove buggy TTL descriptor repair. Previously,
upgrading from 22.2.X to 23.1.9 incorrectly removed TTL storage params from
tables (visible via `SHOW CREATE TABLE <ttl-table>;`) while attempting to
repair table descriptors. This resulted in the node that attempts to run the
TTL job crashing due to a panic caused by the missing TTL storage params.
Clusters currently on 22.2.X should NOT be upgraded to 23.1.9 and should
be upgraded to 23.1.10 or later directly.

ecwall committed:
testing: allow `upgrade all` in logic tests

`upgrade all` upgrades all nodes in the cluster. It has the same behavior as
using `upgrade <idx>` for each index in the cluster.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110500)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
nit: let's add an additional change to this backport. it should use an environment variable to make this function return early by default: https://github.com/cockroachdb/cockroach/blob/f2b6dc7f4cfa4542054d864a09f83a4f3045e0b7/pkg/upgrade/upgrades/first_upgrade.go#L38

rafiss commented on pkg/sql/logictest/testdata/logic_test/mixed_version_upgrade_repair_descriptors:
```diff
@@ -0,0 +1,22 @@
+upgrade all
+
+query B retry
+SELECT version LIKE '%23.1-%' FROM [SHOW CLUSTER SETTING version]
```
in the backport, this should be `LIKE '%22.2-%'`

ecwall committed:
upgrade: add COCKROACH_RUN_FIRST_UPGRADE_PRECONDITION env var

This short-circuits FirstUpgradeFromReleasePrecondition and is false by
default. It has been added to 23.1 releases to prevent automatic descriptor
repair on start up until more test coverage is added.

Release note: None

ecwall reviewed (commented):
Added

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110500#-:-NeBTZUtBIFYMG-7014p:b-8b11j9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_upgrade_repair_descriptors` line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110500#gh-1323554034:-NeBTWXN8jAHCzitzCMa:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf99dead7257dcad36e2e85d5fd5168289b8268/pkg/sql/logictest/testdata/logic_test/mixed_version_upgrade_repair_descriptors#L11)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

in the backport, this should be `LIKE '%22.2-%'`
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice, lgtm!

rafiss commented:
blathers backport 23.1.11-rc

celiala commented:
blathers backport staging-v23.1.10
